helix-core/src/main/java/org/apache/helix/ShardingKeyTrieNode.java

```diff
@@ -0,0 +1,66 @@
+package org.apache.helix;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class ShardingKeyTrieNode {
+  private Map<String, ShardingKeyTrieNode> children;
+  private Map<String, String> data;
```

Comment:
Why is the data field a map vs just a plain string? Or rather, you could just have the `name` field.

Reply:
In the original design, we wanted a leaf node to store a key-value mapping of data, which includes ZkRealm. I'm not sure what a `name` is in this context.

Reply:
What is the original design? You could also publish your rough (or finalized) design in the GitHub wiki if you'd like to refer to it in your PRs.

You _could_ use a map to store data, but it would be good to list out some examples in the form of JavaDoc.

There are pros and cons. If you choose to use one map for all data (for example, name, ZK connect string, etc.), then it'd be hard to require some of these fields (and rightly so - would a TrieNode be valid without a name?). We could consider these items carefully and determine what the right way to represent the data is.

Reply:
Per offline discussion, we are adding two fields, namely `name` and `isTerminal`, and we are changing the `data` field to be a `zkRealm` string field. All fields are final.